letsencrypt/acme/messages.py

```diff
@@ -0,0 +1,491 @@
+            "signature": self.signature,
+        }
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
+        return cls(csr=cls._decode_csr(jobj["csr"]),
+                   signature=other.Signature.from_json(
+                       jobj["signature"], validate=False))
+
+
+@Message.register  # pylint: disable=too-few-public-methods
+class Defer(Message):
+    """ACME "defer" message."""
+    acme_type = "defer"
+    schema = util.load_schema(acme_type)
+    __slots__ = ("token", "interval", "message")
+
+    def _fields_to_json(self):
+        fields = {"token": self.token}
+        if self.interval is not None:
+            fields["interval"] = self.interval
+        if self.message is not None:
+            fields["message"] = self.message
+        return fields
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
+        return cls(token=jobj["token"], interval=jobj.get("interval"),
+                   message=jobj.get("message"))
+
+
+@Message.register  # pylint: disable=too-few-public-methods
+class Error(Message):
+    """ACME "error" message."""
+    acme_type = "error"
+    schema = util.load_schema(acme_type)
+    __slots__ = ("error", "message", "more_info")
+
+    CODES = {
+        "malformed": "The request message was malformed",
+        "unauthorized": "The client lacks sufficient authorization",
+        "serverInternal": "The server experienced an internal error",
+        "notSupported": "The request type is not supported",
+        "unknown": "The server does not recognize an ID/token in the request",
+        "badCSR": "The CSR is unacceptable (e.g., due to a short key)",
+    }
+
+    def _fields_to_json(self):
+        fields = {"error": self.error}
+        if self.message is not None:
+            fields["message"] = self.message
+        if self.more_info is not None:
+            fields["moreInfo"] = self.more_info
+        return fields
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
+        return cls(error=jobj["error"], message=jobj.get("message"),
+                   more_info=jobj.get("moreInfo"))
+
+
+@Message.register  # pylint: disable=too-few-public-methods
+class Revocation(Message):
+    """ACME "revocation" message."""
+    acme_type = "revocation"
+    schema = util.load_schema(acme_type)
+    __slots__ = ()
+
+    def _fields_to_json(self):
+        return {}
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
+        return cls()
+
+
+@Message.register
+class RevocationRequest(Message):
+    """ACME "revocationRequest" message.
+
+    :ivar certificate: Certificate (:class:`M2Crypto.X509.X509`
+        wrapped in :class:`letsencrypt.acme.util.ComparableX509`).
+    :ivar signature: Signature (:class:`letsencrypt.acme.other.Signature`).
+
+    """
+    acme_type = "revocationRequest"
+    schema = util.load_schema(acme_type)
+    __slots__ = ("certificate", "signature")
+
+    @classmethod
+    def create(cls, key, sig_nonce=None, **kwargs):
```

Comment:
That was actually a bug. Everything should be wrapped in `ComparableX509` and I fixed it in recent commits.

`kwargs` accepts whatever class constructor does, and this is all documented by `:ivar:` stanzas. I've added explicit comment about this.